Include Verify certificate rotation docs (#516)
DFE-Digital/dfe-teachers-payment-service: docs/govuk-verify.md

expire after 6 months. See the
[Verify docs for how to rotate keys](https://www.docs.verify.service.gov.uk/maintain-your-connection/rotate-keys/)
for more information.

Pre-requisites:

- You are a requester, see this page on
  [confluence](https://dfedigital.atlassian.net/wiki/spaces/TP/pages/1116110992/Verify+Service+Provider)
- You have access to the Azure management portal for the project
- You have the
  [Azure cli installed](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
- You have the code-base locally

#### Requesting certificates

For non-production (integration) environments (`development`) and production
environments (`production`), we need to request certificates from Verify. The
following will guide you though the process of generating, enrolling and
rotating certificates.

##### Create certificate signing requests (CSR)

Complete these steps for both signing and encryption certificates.

Access the Azure portal and locate the key vault for the environment you are
updating the certificates for

https://azure.microsoft.com/

Note the version number of the latest set of certificates, keys, CSRs and

> challenge phrase e.g. TeacherPayments`ENV`VspSaml`TYPEN`, where:

Reviewer: That `TYPEEN` is difficult to parse. I prefer to do examples with curly brackets like this:

```
TeacherPayments{env}VspSaml{type}{version_number}
```

Might be an idea if we use a more descriptive name for the version number too

- `ENV` is the environment (either Dev or Prod)
- `TYPE` is the type of certificate (Signing or Encryption)
- `N` is the version number

Run the request script with the next version number in the sequence:

```bash
bin/request-vsp-certs <environment> <version>
```

Generate a strong challenge phrase and keep a note of it as you will need it
during enrolment (see below), the challenge phrase should contain only
alphabetical or numerical characters. The following special characters are also

> allowed `' ( ) + , - . / ,` and `space`.

Reviewer: Can we seperarte out each character in its own backtick? I stumbled over this a bit

Once complete, the csr, keys and pass-phrase are all added to your local
directory in `tmp/certs` and uploaded to the Azure Keyvault.

**Be mindful not to commit these files and delete them once your are confident

> they have been enrolled (see below).**

Reviewer: Do we `.gitignore` this directory?

In local development mode, nothing needs to be done.

### Live environments

Integration certificates expire after 2 years, while production certificates
expire after 6 months. See the
[Verify docs for how to rotate keys](https://www.docs.verify.service.gov.uk/maintain-your-connection/rotate-keys/)
for more information.

Pre-requisites:

- You are a requester, see this page on
  [confluence](https://dfedigital.atlassian.net/wiki/spaces/TP/pages/1116110992/Verify+Service+Provider)
- You have access to the Azure management portal for the project
- You have the
  [Azure cli installed](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
- You have the code-base locally

#### Requesting certificates

For non-production (integration) environments (`development`) and production
environments (`production`), we need to request certificates from Verify. The
following will guide you though the process of generating, enrolling and
rotating certificates.

##### Create certificate signing requests (CSR)

Complete these steps for both signing and encryption certificates.

Access the Azure portal and locate the key vault for the environment you are

> updating the certificates for

Reviewer 2: ```suggestion
updating the certificates for.
```

> https://azure.microsoft.com/

Reviewer 2: We should probably be linking to https://portal.azure.com instead.

Note the version number of the latest set of certificates, keys, CSRs and

> challenge phrase e.g. TeacherPayments`{ENV}`VspSaml`{TYPE}` `{VERSION_NUMBER}`,

Reviewer 2: ```suggestion
challenge phrase e.g. TeacherPayments`{ENV}`VspSaml`{TYPE}``{VERSION_NUMBER}`,
```

where:

- `ENV` is the environment (either Dev or Prod)
- `TYPE` is the type of certificate (Signing or Encryption)
- `VERSION_NUMBER` is the version number

Run the request script with the next version number in the sequence:

```bash
bin/request-vsp-certs <environment> <version>
```

Generate a strong challenge phrase and keep a note of it as you will need it
during enrolment (see below), the challenge phrase should contain only
alphabetical or numerical characters. The following special characters are also
allowed `'` `(` `)` `+` `,` `-` `.` `/` `,` and `space`.

> Once complete, the csr, keys and pass-phrase are all added to your local

Reviewer 2: ```suggestion
Once complete, the CSR, keys and challenge phrase are all added to your local
```

directory in `tmp/certs` and uploaded to the Azure Keyvault.

> **Be mindful to delete these files once your are confident they have been

Reviewer 2: ```suggestion
**Be mindful to delete these files once you are confident they have been
```

enrolled (see below).**

##### Enrol

Complete these steps for both signing and encryption certificates.

Open the following document:

> https://docs.google.com/document/d/1lH4b2uOn35rdzLAzW5I-_O9wG9NU8fTre8FaM0Y97L8/edit?pli=1#

Reviewer 2: ```suggestion
<https://docs.google.com/document/d/1lH4b2uOn35rdzLAzW5I-_O9wG9NU8fTre8FaM0Y97L8/edit?pli=1#>
```

This tells the Markdown that it's a link, and stops the `_` being parsed.

> https://docs.google.com/document/d/1lH4b2uOn35rdzLAzW5I-_O9wG9NU8fTre8FaM0Y97L8/edit?pli=1#

Reviewer 2: Are ok to link to this publicly in the repo? We got shared it in private.

And go to the appropriate url for the environment you are enrolling the
certificates for, either integration or production.

- Choose ENROL
- Upload the CSR file
- Enter the details in the form:
  - Applicant Details: your name and the group email address

>     (capt-dev@digital.education.gov.uk), ensue the email address is correct as

Reviewer 2: ```suggestion
    (capt-dev@digital.education.gov.uk), ensure the email address is correct as
```

    this is where the certificate will be sent.
  - Certificate Profile: Select the type of certificate either SAML Signing or
    SAML Encryption, ensure this matches the CSR you are uploading
  - Challenge Phrase: the challenge phrase used above

Once submitted you will receive a confirmation email, the certificates will be
generated within 5 days and we be sent to the group email address.

##### Import

When you receive the certificates run the import script for both signing and
encryption certificates:

```bash
bin/import-vsp-certs <environment> <version>
```

Using the same environment and version as you set in the `request-vsp-certs`
script.

When prompted, paste the certificate body for the type of certificate (signing
or encryption) - be careful to note which type of certificate you are importing.

The script copies the certificates to the Azure Keyvault.

##### Rotate

The steps to actually rotate the keys varies between the two types of

> certificate

Reviewer 2: ```suggestion
certificate. See the
[Verify docs for this step](https://www.docs.verify.service.gov.uk/maintain-your-connection/rotate-keys/#rotate-your-encryption-and-signing-keys).
```

[Verify docs for this step](https://www.docs.verify.service.gov.uk/maintain-your-connection/rotate-keys/#rotate-your-encryption-and-signing-keys)

> ###### Signing

Reviewer 2: ```suggestion
###### Signing key
```

It think it would be good to be clear that this isn't an action "we need to sign something" and is a label of the type of key.

Now that the signing key is stored in the Azure Keyvault, the next step gets the
application to use it:

- Send the GOVUK Verify team an email to
  idasupport+onboarding@digital.cabinet-office.gov.uk with the signing
  certificate letting them know you are rotating the signing key

> Once you have a response that the certificate is deployed, move on

Reviewer 2: ```suggestion
Once you have a response that the certificate is deployed:
```

> - Locate the `production.template.json` and/or `development.templtate.json` in

Reviewer 2: ```suggestion
- Locate the `production.template.json` and/or `development.template.json` in
```

  `azure/app/parameters` for the environment you are updating
- Open the file and edit the `VSP.SAML_SIGNING_KEY` entry
- You'll see the `secret_name` matches the one in the Azure Keyvault with the
  old version number
- Update the version number, commit the changes and open a PR to get them merged

> - The next deploy will update the keys, if updating production you will have to

Reviewer 2: ```suggestion
- The next deploy will update the keys. If updating production you will have to
```

  perform a release

> - Let the GOVUK Verify team know the signing key has been rotated, they will

Reviewer 2: ```suggestion
- Let the GOVUK Verify team know the signing key has been rotated, and they will
```

  delete the old key

> ###### Encryption

Reviewer 2: ```suggestion
###### Encryption key
```

It think it would be good to be clear that this isn't an action "we need to encrypt something" and is a label of the type of key.

Now that the encryption key is stored in the Azure Keyvault, the next step gets
the application to use it:

> - Locate the `production.template.json` and/or `development.templtate.json` in

Reviewer 2: ```suggestion
- Locate the `production.template.json` and/or `development.template.json` in
```

>   azure/app/parameters for th`e environment you are updating

Reviewer 2: ```suggestion
  `azure/app/parameters` for the environment you are updating
```

> - Open the file and edit the VSP.SAML_SECONDARY_ENCRYPTION_KEY` entry

Reviewer 2: ```suggestion
- Open the file and edit the `VSP.SAML_SECONDARY_ENCRYPTION_KEY` entry
```

- The `VSP.SAML_SECONDARY_ENCRYPTION_KEY` entry will have a `value` of an empty
  string

> - Create a new object in VSP.SAML_SECONDARY_ENCRYPTION_KEY that references your

Reviewer 2: ```suggestion
- Create a new object in `VSP.SAML_SECONDARY_ENCRYPTION_KEY` that references your
```

  new certificate, for example:

```json
"VSP.SAML_SECONDARY_ENCRYPTION_KEY": {
  "reference": {
    "keyVault": {
      "id": "${keyVaultId}"
    },
    "secretName": "TeacherPaymentsProdVspSamlEncryption3KeyBase64"
  }
},

> ```

Reviewer 2: Indent these to the same level as the bullet to nest them in the bullet list.

> - Commit, review and merge this change, if updating production you will have to

Reviewer 2: ```suggestion
- Commit, review and merge this change. If updating production you will have to
```

  perform a release

At this point the service is able to use both the old and the new encryption
keys.

- Send the GOVUK Verify team an email to
  idasupport+onboarding@digital.cabinet-office.gov.uk with the encryption
  certificate letting them know we are rotating the encryption key

Once you have a response that the certificate is deployed, move on

- Move the object from `VSP.SAML_SECONDARY_ENCRYPTION_KEY` to
  `VSP.SAML_PRIMARY_ENCRYPTION_KEY`

> - Set the VSP.SAML_SECONDARY_ENCRYPTION_KEY value to an empty string:

Reviewer 2: ```suggestion
- Set the `VSP.SAML_SECONDARY_ENCRYPTION_KEY` value to an empty string:
```

```json
    "VSP.SAML_SECONDARY_ENCRYPTION_KEY": {
    "value": ""
    },

> ```

Reviewer 2: Indent these to the same level as the bullet to nest them in the bullet list.

> - commit, review and merge this change, if updating production you will have to

Reviewer 2: ```suggestion
- Commit, review and merge this change. If updating production you will have to
```

  perform a release

#### Clean up

> Remove any copies of the files from `tmp/certs`.

Reviewer 2: ```suggestion
Confirm the Keyvault contains all of the generated keys and secrets. If so, remove any copies of the files from `tmp/certs`.
```
